pkg/tbtc/deduplicator.go

```diff
@@ -55,3 +62,25 @@ func (d *deduplicator) notifyDKGStarted(
 	// with the execution.
 	return false
 }
+
+// notifyDKGResultSubmitted notifies the client wants to start some actions
+// upon the DKG result submission. It returns boolean indicating whether the
+// client should proceed with the actions or ignore the event as a duplicate.
+func (d *deduplicator) notifyDKGResultSubmitted(
+	newDKGResultHash [32]byte,
+) bool {
+	d.dkgResultHashCache.Sweep()
+
+	// The cache key is the hexadecimal representation of the hash.
+	cacheKey := hex.EncodeToString(newDKGResultHash[:])
+	// If the key is not in the cache, that means the result was not handled
+	// yet and the client should proceed with the execution.
+	if !d.dkgResultHashCache.Has(cacheKey) {
+		d.dkgResultHashCache.Add(cacheKey)
+		return true
+	}
+
+	// Otherwise, the DKG result is a duplicate and the client should not
+	// proceed with the execution.
+	return false
+}
```

Comment:
I think we'll have a security hole if we do it this way. I am looking at `EcdsaDkg.sol` and nothing in `submitResult` is blocking me from submitting the same result after it's been challenged. Consider this scenario:
1. Mallory is submitting malicious result.
2. All nodes challenge it, Mallory gets slashed.
3. Mallory is submitting the same result one more time.
4. None of the nodes care about it because they reject the event as a duplicate.

We could be checking if the event comes from the same block as the previous one for the given seed. If yes, reject as a duplicate. But I think this will still be vulnerable in this scenario:
1. Mallory is submitting malicious result at block N.
2. The node receives an event and challenges the result at block N+1.
3. There is soft work and the chain gets reorganized so that block N+1 is rejected but block N stays.

I would consider not looking at the events at all for challenges and instead, periodically check the state of the chain. If we are in the challenge period, check the result.

Reply:
I'm afraid we need to use events anyway. A periodic check of the DKG state is not an option as you need to have a plain-text result that was submitted. The contract stores just the hash. I think we have to shorten the cache time to a reasonable value, that will exclude duplicated events on the one hand, and prevent the described scenario on the other hand. WDYT?

Reply:
We can use `PastDkgResultSubmittedEvents` exposed by the `WalletRegistry` contract Go bindings to obtain the last event if we decide for the periodic checks option.

I am not sure if there is a *reasonable value* for the cache time in this case. It would depend too much on the current situation: what is the block time, did we have a chain reorg, when the malicious result got re-submitted.

I think we have two options:
1. Do the check periodically, use `PastDkgResultSubmittedEvents`
2. Accept we may receive duplicates for `DkgResultSubmittedEvent` and always do the validation.

I think the first option is cleaner and safer: all nodes, all the time, watch the situation. If they detect a problem, they all try to act. This is like a watchman process in the background that never sleeps.

Reply:
I think `PastDkgResultSubmittedEvents` will also provide us with duplicates. This is often the case when the client needs to reconnect with their ETH node. That said, it seems this will not improve our situation. My personal preference is to do the following:
- Always validate without checking duplicates and challenge if the result is invalid
- If the result is valid, check if the event is not duplicated and if not, schedule the result approval 

What do you think?

Reply:
Option 2 turned out to be much more complicated than it seemed:
- We don't have a simple way to determine an efficient block span and the current seed to pre-filter fetched events. That may cause problems with some ETH nodes that limit the number of returned past events. Determining proper filter values requires a separate mechanism that greatly increases the complexity of the entire solution
- We need to have a mechanism that makes sure we schedule the approval for the given result only once
- A lot of calls to the ETH node must be made
- We run the check every 30 minutes while DKG is much more sparse

I finally came up with option 1 with some modifications to the deduplication logic (see b70ccb8b). The deduplicator now considers the DKG seed and the submission block for the given result. This way, we can still validate the same malicious result submitted multiple times, even for different DKGs. This idea comes from https://github.com/keep-network/keep-core/pull/3427#discussion_r1053192176 and, as mentioned, is still vulnerable to chain reorgs. However, this problem should be mitigated by a challenge confirmation mechanism that was introduced according to the original option 1 description. That said, the solution presented in b70ccb8b should address all problems mentioned in this discussion. Even though the client loses the challenge confirmation state upon restart, that should not be a problem as we have redundancy coming from other clients working in the network that should also take care of the challenge. In my opinion, the presented solution is also an optimal one regarding code quality. The original option 1 leads to a weird structure when the result deduplication logic is mixed with the DKG executor code.